### PR TITLE
GH-1334 AMCHART_BACKWARD_COMPAT

### PR DESCRIPTION
**Pull request by mlzummo**
Adds PP for backwards compat and old publish params back

Fixes GH-1334

Signed-off-by: Mathew Zummo <mzummo@gmail.com>

### CODE REVIEW

**Comment by mlzummo** [comment]
need to figure out how how to hide the new Axis Obj from the property editor (prolly going to use new PE excludes and have the PE read in what needs to be exlcuded from the chart itself)

**Comment by mlzummo** [comment]
@GordonSmith i think im going to release this as a 1.10.X fix if jo P has any issues since he uses these charts alot i have property editor changes that need tob e made as welll

File: src/amchart/CommonSerial.js [hunk 1]
```diff
@@ -31,6 +31,8 @@
     CommonSerial.prototype.constructor = CommonSerial;
     CommonSerial.prototype._class += " amchart_CommonSerial";
 
+    CommonSerial.prototype.publish("backwardsCompatible", true, "boolean", "Allow use of old publish parameters");
+
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: Remove - for the 1.x releases this should always be true and should not be an option.

File: src/amchart/CommonSerial.js [hunk 2]
```diff
@@ -59,7 +61,7 @@
 
     CommonSerial.prototype.publish("useClonedPalette", false, "boolean", "Enable or disable using a cloned palette",null,{tags:["Intermediate","Shared"]});
 
-    CommonSerial.prototype.publish("yAxisTickFormat", "s", "string", "Y-Axis Tick Format");
+    //CommonSerial.prototype.publish("yAxisTickFormat", "s", "string", "Y-Axis Tick Format");
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: remove commented code

File: src/amchart/CommonSerial.js [hunk 3]
```diff
@@ -80,6 +82,67 @@
 
     CommonSerial.prototype.publish("y2", [], "array", "Columns to associate with second Y-Axis");
 
+//
+
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: Remove commented code  and blank lines

File: src/amchart/CommonSerial.js [hunk 4]
```diff
@@ -80,6 +82,67 @@
...
+    CommonSerial.prototype.publish("xAxisGridAlpha", 0.0, "number", "Grid alpha.",null,{tags:["Intermediate"]});
+    CommonSerial.prototype.publish("yAxisGridAlpha", 0.0, "number", "Grid alpha.",null,{tags:["Intermediate"]});
+
+    //CommonSerial.prototype.publish("yAxisMinimum", null, "number", "",null,{tags:["Intermediate"]});
+    CommonSerial.prototype.publish("yAxisTitleOffset", null, "number", "",null,{tags:["Intermediate"]});
+
+//
+
+
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: Remove commented code and blank lines

File: src/amchart/CommonSerial.js [hunk 5]
```diff
@@ -80,6 +82,67 @@
...
+    CommonSerial.prototype.publish("yAxisFillAlpha", 0, "number", "Fill opacity. Every second space between grid lines can be filled with color. Set fillAlpha to a value greater than 0 to see the fills.",null,{tags:["Intermediate"]});
+
+    CommonSerial.prototype.publish("xAxisFillColor", null, "html-color", "Fill color. Every second space between grid lines can be filled with color. Set fillAlpha to a value greater than 0 to see the fills.",null,{tags:["Intermediate"]});
+    CommonSerial.prototype.publish("yAxisFillColor", null, "html-color", "Fill color. Every second space between grid lines can be filled with color. Set fillAlpha to a value greater than 0 to see the fills.",null,{tags:["Intermediate"]});
+
+    CommonSerial.prototype.publish("xAxisGridAlpha", 0.0, "number", "Grid alpha.",null,{tags:["Intermediate"]});
+    CommonSerial.prototype.publish("yAxisGridAlpha", 0.0, "number", "Grid alpha.",null,{tags:["Intermediate"]});
+
+    //CommonSerial.prototype.publish("yAxisMinimum", null, "number", "",null,{tags:["Intermediate"]});
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: Remove commented code and blank lines

File: src/amchart/CommonSerial.js [hunk 6]
```diff
@@ -210,135 +274,216 @@
...
+        this._chart.categoryAxis.dashLength = (this.xAxisDashLength() && this.backwardsCompatible()) ? this.xAxisDashLength() : xAxis.axisDashLength();
+        this._chart.categoryAxis.fillAlpha = (this.xAxisFillAlpha() && this.backwardsCompatible()) ? this.xAxisFillAlpha() : xAxis.axisFillAlpha();
+        this._chart.categoryAxis.fillColor = (this.xAxisFillColor() && this.backwardsCompatible()) ? this.xAxisFillColor() : xAxis.axisFillColor();
+        this._chart.categoryAxis.fontSize = (this.axisFontSize() && this.backwardsCompatible()) ? this.axisFontSize() : xAxis.axisFontSize();
+        this._chart.categoryAxis.color = (this.xAxisFontColor() && this.backwardsCompatible()) ? this.xAxisFontColor() : xAxis.axisFontColor();
+        this._chart.categoryAxis.titleColor = (this.xAxisTitleFontColor() && this.backwardsCompatible()) ? this.xAxisTitleFontColor() : xAxis.axisTitleFontColor();
+        this._chart.categoryAxis.titleFontSize = (this.xAxisTitleFontSize() && this.backwardsCompatible()) ? this.xAxisTitleFontSize() : xAxis.axisTitleFontSize();
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: Remove the backwardsCompatible flag and assume it to be true.

File: src/amchart/CommonSerial.js [hunk 7]
```diff
@@ -210,135 +274,216 @@
...
-                this._chart.categoryAxis.parseDates = false;
-                this._chart.categoryAxis.logarithmic = false;
-                this.dataFormatter = xAxis.axisTickFormat() ? d3.format(xAxis.axisTickFormat()) : function(v) { return v; };
-                break;
-        }
-
-        var context = this;
-        this._chart.categoryAxis.labelFunction = function(v1, v2, v3) {
-            switch (xAxis.axisType()) {
+        if (this.backwardsCompatible()) {
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: Remove this and assume its true.

File: src/amchart/CommonSerial.js [hunk 8]
```diff
@@ -210,135 +274,216 @@
...
-                        return context.valueFormatter(v1);
+                        return context.dataFormatter(v1);
                 }
             };
         }
+        for (var i = 0; i < this.yAxes().length; i++) {
+            var yAxis = this.yAxis()[i];
+            yAxis.type("y");
+
+            if (!this._chart.valueAxes[i] && !this.backwardsCompatible()) {
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: This if is not needed.

File: src/amchart/CommonSerial.js [hunk 9]
```diff
@@ -210,135 +274,216 @@
...
+            this._chart.valueAxes[i].fillColor = (this.yAxisFillColor() && this.backwardsCompatible()) ? this.yAxisFillColor() : yAxis.axisFillColor();
+            this._chart.valueAxes[i].fillAlpha = (this.yAxisFillAlpha() && this.backwardsCompatible()) ? this.yAxisFillAlpha() : yAxis.axisFillAlpha();
+            this._chart.valueAxes[i].labelRotation = (this.xAxisLabelRotation() && this.backwardsCompatible()) ? this.xAxisLabelRotation() : yAxis.axisLabelRotation(); // need to check serial axis object
+
+            this._chart.valueAxes[i].gridAlpha = (this.yAxisGridAlpha() && this.backwardsCompatible()) ? this.yAxisGridAlpha() : yAxis.axisGridAlpha();
+            this._chart.valueAxes[i].dashLength = (this.yAxisDashLength() && this.backwardsCompatible()) ? this.yAxisDashLength() : yAxis.axisDashLength();
+            this._chart.valueAxes[i].boldPeriodBeginning = (this.yAxisBoldPeriodBeginning() && this.backwardsCompatible()) ? this.yAxisBoldPeriodBeginning() : yAxis.axisBoldPeriodBeginning();
+            //this._chart.valueAxes[0].axisTitleOffset = this.yAxisTitleOffset(); ??
+            if (this.backwardsCompatible()) {
+                switch(this.yAxisType()) {
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: Remove the backwardsCompatible flag and assume it to be true.

File: src/amchart/CommonSerial.js [hunk 10]
```diff
@@ -358,10 +503,18 @@
             this._chart.chartCursor.categoryBalloonEnabled = false;
         }
 
-        this._currXAxisTypes = this.xAxes().map(function(axe) { return axe.axisType(); }).toString();
-        this._currYAxisTypes = this.yAxes().map(function(axe) { return axe.axisType(); }).toString();
-        this._currXAxisTypeTimePatterns = this.xAxes().map(function(axe) { return axe.axisTypeTimePattern(); }).toString();
-        this._currYAxisTypeTimePatterns = this.yAxes().map(function(axe) { return axe.axisTypeTimePattern(); }).toString();
+        if (this.backwardsCompatible()) {
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: Remove the backwardsCompatible flag and assume it to be true.

File: src/amchart/CommonXY.js [hunk 11]
```diff
@@ -24,6 +24,8 @@
     CommonXY.prototype.constructor = CommonXY;
     CommonXY.prototype._class += " amchart_CommonXY";
 
+    CommonXY.prototype.publish("backwardsCompatible", true, "boolean", "Allow use of old publish parameters");
```
Comment (GordonSmith) [inline src/amchart/CommonXY.js]: Remove the backwardsCompatible flag and assume it to be true.

File: src/amchart/CommonXY.js [hunk 12]
```diff
@@ -111,24 +160,29 @@
...
+            this._chart.valueAxes[vAxisCount].autoGridCount = (this.yAxisAutoGridCount() && this.backwardsCompatible()) ? this.yAxisAutoGridCount() : yAxis.axisAutoGridCount();
+            this._chart.valueAxes[vAxisCount].gridPosition = (this.yAxisGridPosition() && this.backwardsCompatible())  ? this.yAxisGridPosition() : yAxis.axisGridPosition();
+            this._chart.valueAxes[vAxisCount].fillAlpha = (this.yAxisFillAlpha() && this.backwardsCompatible()) ? this.yAxisFillAlpha() : yAxis.axisFillAlpha();
+            this._chart.valueAxes[vAxisCount].fillColor = (this.yAxisFillColor() && this.backwardsCompatible()) ? this.yAxisFillColor() : yAxis.axisFillColor();
+            this._chart.valueAxes[vAxisCount].gridAlpha = (this.yAxisGridAlpha() && this.backwardsCompatible()) ? this.yAxisGridAlpha() : yAxis.axisGridAlpha();
+            this._chart.valueAxes[vAxisCount].dashLength = (this.yAxisDashLength() && this.backwardsCompatible()) ? this.yAxisDashLength() : yAxis.axisDashLength();
+
+            if (this.backwardsCompatible()) {
```
Comment (GordonSmith) [inline src/amchart/CommonXY.js]: Remove the backwardsCompatible flag and assume it to be true.

File: src/amchart/CommonSerial.js (hunk 6 above)
Comment (mlzummo) [inline src/amchart/CommonSerial.js]: not going to happen ... might as well just revert to the old code base then and revove the idea of the axis object

**Comment by mlzummo** [comment]
@GordonSmith the @backwardsCompatible  is there to stay because there has to be a way to switch between using the old params and the new ones (with both having default values assigned by amacharts themselves)

**Comment by mlzummo** [comment]
@GordonSmith i need to figure out a way to hide the axis obj params when backwards compat is on (so only the old ones show and vise versa .... any ideas)

**Comment by GordonSmith** [comment]
For the 1.x branch you can't hide the legacy PP - and if you don't want to show the new ones, delete them.

**Comment by mlzummo** [comment]
@GordonSmith please re-review ... i still have the backwards compat in there but take a look at how i have it all working (i think i still have 1 more thing to do to hide the old params when the new ones are selected)

http://rawgit.com/mlzummo/Visualization/AMCHART_BACKWARD_COMPAT/demos/dermatology.html?src/amchart/Column.simple

File: src/amchart/CommonSerial.js (hunk 7 above)
Comment (mlzummo) [inline src/amchart/CommonSerial.js]: cant because the logic changes below it

File: src/amchart/CommonSerial.js [hunk 13]
```diff
@@ -80,6 +88,81 @@
 
     CommonSerial.prototype.publish("y2", [], "array", "Columns to associate with second Y-Axis");
 
+//
+
+    CommonSerial.prototype.publish("axisFontSize", null, "number", "X/Y Axis Text Font Size",null,{tags:["Basic","Shared"]});
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: For the ones with a 1 to 1 mappings could you simply use:
```
CommonSerial.prototype.publishProxy("axisFontSize", "xAxis", "axisTitle");
```

File: src/amchart/CommonSerial.js [hunk 14]
```diff
@@ -80,6 +88,81 @@
...
+
+    CommonSerial.prototype.publish("xAxisBaselineColor", "#000000", "html-color", "X Axis Baseline Color",null,{tags:["Basic","Shared"]});
+    CommonSerial.prototype.publish("yAxisBaselineColor", "#000000", "html-color", "Y Axis baseline Color",null,{tags:["Basic","Shared"]});
+
+    CommonSerial.prototype.publish("xAxisFontColor", null, "html-color", "Horizontal Axis Text Style (Color)",null,{tags:["Basic","Shared"]});
+    CommonSerial.prototype.publish("yAxisFontColor", null, "html-color", "Vertical Axis Text Style (Color)",null,{tags:["Basic","Shared"]});
+
+    CommonSerial.prototype.publish("xAxisTitle", "", "string", "X-Axis Title",null,{tags:["Basic","Shared"]});
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: For the ones with a 1 to 1 mappings could you simply use:
```
CommonSerial.prototype.publishProxy("xAxisTitle", "xAxis", "axisTitle");
```

**Comment by mlzummo** [comment]
going to close in favor of : https://github.com/hpcc-systems/Visualization/pull/1374

**Comment by mlzummo** [comment]
@GordonSmith make sure to close this in favor of the other branch

**Comment by GordonSmith** [comment]
Closing in favor of:  https://github.com/hpcc-systems/Visualization/pull/1377

**Pull request closed without merging**